### PR TITLE
#9609: Reorganize libs into ttnn

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
#9609

### Problem description
We are merging tt_eager and ttnn into one.
We approach it like this:
* Create symlink. Remove tt_lib.so, merge tt_lib bindings into ttnn, merge tt_dnn and the rest of of tt_eager libs into ttnn
* Wait till massive ops migration slows down
* Remove tt_eager folder completely, remove symlink. Put files inside ttnn

### What's changed

- Added symlink tt_eager --> ttnn/cpp/ttnn/experimental
- Removed tt_lib.so
- Removed tt_eager libraries from cmake
- Restructured ttnn cmake to build a single lib from ttnn + tt_eager sources
- ttnn.so now contains both tt_lib and ttnn bindings 

**🪣 Changes can be grouped into:**
- CMake changes
- Packaging changes
- Bindings changes
- Py modules changes to ttnn/tt_lib
- Py code adjustments of imports and namespaces
- ttnn self-reliance (as much as possible eliminate dependency on import tt_lib in ttnn)
- Blocking fix, which ideally one merges to main separately, but I wanted to avoid switching branches
- Clueless changes (I hope there is a minimal amount of that)

Future changes:
- We can build metal as static library
- Replace tt_lib usage across whole codebase with ttnn. Remove tt_lib
- Reorganize ttnn to find proper placement for old tt_lib tensor/device/profiler/operations..

### Checklist
- [x] Post commit CI [passed](https://github.com/tenstorrent/tt-metal/actions/runs/9899387467/job/27348190919)
- [x] t3k Frequent tests [no regression](https://github.com/tenstorrent/tt-metal/actions/runs/9877565252/job/27279921193 )
- [x] t3k model perf test [no regression](https://github.com/tenstorrent/tt-metal/actions/runs/9882566495)
- [x] t3 demo tests [no regression](https://github.com/tenstorrent/tt-metal/actions/runs/9882562999)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp [hunk 1]
```diff
@@ -683,7 +687,7 @@ std::function<std::vector<ttnn::Tensor>(const Tensor&, const Tensor&, const Tens
     }
 }
 
-std::function<std::vector<ttnn::Tensor>(const Tensor&, const Tensor&, const Tensor&, string, const MemoryConfig&)> get_function_type1_w_string(BinaryBackwardOpType OpType){
+std::function<std::vector<ttnn::Tensor>(const Tensor&, const Tensor&, const Tensor&, std::string, const MemoryConfig&)> get_function_type1_w_string(BinaryBackwardOpType OpType){
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp:690]: Please read my comments in the https://github.com/tenstorrent/tt-metal/pull/10037.
Also why do we pass std::string by value. Probably string_view or at least  const std::string& might be better.

**Review by dmakoviichuk-tt — COMMENTED** [review]
As I see your goal is not to add new functionality but rearrange libraries and remove tt_eager. Probably you don't need to do any huge changes in code logic  in the same pr.
Feel free to ignore my comment here. It should be separate pr.

File: ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp (hunk 1 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary_backward/device/binary_backward_op.cpp:690]: 100%
This whole file is on fire. I am not touching it in this PR on purpose, but had to do something to unblock w/o switching branches.

File: models/demos/resnet/tt/metalResnetBlock50.py [hunk 2]
```diff
@@ -23,11 +23,11 @@
     resnet50_optimized_conv,
     resnet50_1x1_conv_s2_as_downsample_and_matmul,
 )
-from tt_eager.tt_dnn.op_library.sliding_window_op_infra.tt_py_composite_conv import (
```
Comment (ayerofieiev-tt) [inline models/demos/resnet/tt/metalResnetBlock50.py:26]: no more `from tt_eager.` anywhere else in the codebase

**Review by eyonland — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by boris-drazic — APPROVED** [review]

File: create_venv.sh [hunk 3]
```diff
@@ -34,7 +34,6 @@ python3 -m pip install -r $(pwd)/tt_metal/python_env/requirements-dev.txt
 
 echo "Installing tt-metal"
 pip install -e .
-pip install -e ttnn
```
Comment (tt-rkim) [inline create_venv.sh:37]: <3 <3 <3

File: models/demos/t3000/falcon40b/scripts/get_falcon40b_perf_numbers.sh [hunk 4]
```diff
@@ -12,7 +12,7 @@ dtype="BFLOAT8_B" #BFLOAT16
 for seq_len in ${seq_lens[@]}; do
     echo "Running seq_len: $seq_len"
     output_folder="generated/profiler/reports/"
-    WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml python3 tt_eager/tracy.py -r -m "pytest models/demos/t3000/falcon40b/tests/test_perf_falcon.py::test_perf_bare_metal[${dtype}-DRAM-falcon_40b-layers_1-prefill_seq${seq_len}-8chips]"
+    WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml python3 -m tracy -r -m "pytest models/demos/t3000/falcon40b/tests/test_perf_falcon.py::test_perf_bare_metal[${dtype}-DRAM-falcon_40b-layers_1-prefill_seq${seq_len}-8chips]"
```
Comment (tt-rkim) [inline models/demos/t3000/falcon40b/scripts/get_falcon40b_perf_numbers.sh:15]: <3

File: ttnn/CMakeLists.txt [hunk 5]
```diff
@@ -61,23 +68,33 @@ target_precompile_headers(ttnn_lib PRIVATE
     <vector>
 )
 
-set(TTNN_FINAL_SRC $<TARGET_OBJECTS:ttnn_lib>)
-set(TTNN_PUBLIC_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/cpp)
-set(TTNN_PUBLIC_LINK_LIBS compiler_flags linker_flags tt_eager) # linker_flags = -rdynamic if tracy enabled
+### Setup TTNN as a shared library with optional Python bindings
+set(TTNN_FINAL_SRC $<TARGET_OBJECTS:ttnn_lib> ${QUEUE_SRCS} ${TENSOR_SRCS} ${TT_DNN_SRCS})# $<TARGET_OBJECTS:tt_dnn> $<TARGET_OBJECTS:tensor> $<TARGET_OBJECTS:queue>)
```
Comment (tt-rkim) [inline ttnn/CMakeLists.txt:72]: Do we need to keep this comment appended here?

File: ttnn/cpp/pybind11/__init__.cpp [hunk 6]
```diff
@@ -11,13 +11,17 @@
...
 
+#include "ttnn/experimental/tt_lib/csrc/tt_lib_bindings.hpp"
 #include "operations/__init__.hpp"
 
 namespace py = pybind11;
 
 PYBIND11_MODULE(_ttnn, module) {
     module.doc() = "Python bindings for TTNN";
 
+    auto m_deprecated = module.def_submodule("deprecated", "tt_lib deprecated bindigns tensor, device, profiler..");
```
Comment (tt-rkim) [inline ttnn/cpp/pybind11/__init__.cpp:22]: spelling

Comment (tt-rkim) [reply ttnn/cpp/pybind11/__init__.cpp:22]: pedantic because it seems like an important import for ppl

File: ttnn/CMakeLists.txt [hunk 7]
```diff
@@ -37,20 +37,27 @@ set(TTNN_SRCS
...
 
-add_library(ttnn_lib OBJECT ${TTNN_SRCS})
-target_compile_options(ttnn_lib PUBLIC -MP -Wno-int-to-pointer-cast -fno-var-tracking)
-target_link_libraries(ttnn_lib
-    PUBLIC compiler_flags metal_header_directories metal_common_libs
-)
-target_include_directories(ttnn_lib PUBLIC
+### Setup TTNN_LIB as an Object library
+# Save variables to reuse between ttnn_lib and ttnn
+# Have to manually propagate because ttnn_lib is an OBJECT library
```
Comment (tt-rkim) [inline ttnn/CMakeLists.txt:42]: Really? Object libraries don't propagate this stuff through their public interface? Didn't know that!

File: ttnn/tracy/__main__.py
Comment (tt-rkim) [inline ttnn/tracy/__main__.py:1]: SO MUCH BETTER THIS WAYYY THANK YOUUU
@mo-tenstorrent artem cleaned it up for us what a legend

**Review by tt-rkim — APPROVED** [review]

File: ttnn/CMakeLists.txt (hunk 7 above)
Comment (ayerofieiev-tt) [reply ttnn/CMakeLists.txt:42]: Yes. They do not propagate up from their own dependencies. Very annoying.

File: ttnn/CMakeLists.txt (hunk 5 above)
Comment (ayerofieiev-tt) [reply ttnn/CMakeLists.txt:72]: Thanks!

File: ttnn/cpp/pybind11/__init__.cpp (hunk 6 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/__init__.cpp]: Thank you!

File: ttnn/tracy/__init__.py [hunk 8]
```diff
@@ -188,7 +188,7 @@ def split_comma_list(option, opt, value, parser):
 def main():
     from optparse import OptionParser
 
-    usage = "python -m tracy.py [-m module | scriptfile] [arg] ..."
```
Comment (mo-tenstorrent) [inline ttnn/tracy/__init__.py:191]: @ayerofieiev-tt Shoot I see it now. Thanks man

File: ttnn/tracy/__main__.py
Comment (mo-tenstorrent) [reply ttnn/tracy/__main__.py:1]: Legend indeed ❤️

**Pull request merged**